components/fxa-client/src/commands/send_tab.rs

```diff
@@ -0,0 +1,192 @@
+    pub fn decrypt(self, keys: &PrivateSendTabKeys) -> Result<SendTabPayload> {
+        let encrypted = base64::decode_config(&self.encrypted, base64::URL_SAFE_NO_PAD)?;
+        let private_key = LocalKeyPairImpl::new(&keys.private_key)?;
+        let decrypted =
+            Aes128GcmEceWebPushImpl::decrypt(&private_key, &keys.auth_secret, &encrypted)?;
+        Ok(serde_json::from_slice(&decrypted)?)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct SendTabPayload {
+    pub entries: Vec<TabData>,
+}
+
+impl SendTabPayload {
+    pub fn single_tab(title: &str, url: &str) -> Self {
+        SendTabPayload {
+            entries: vec![TabData {
+                title: title.to_string(),
+                url: url.to_string(),
+            }],
+        }
+    }
+    fn encrypt(&self, keys: PublicSendTabKeys) -> Result<EncryptedSendTabPayload> {
+        let bytes = serde_json::to_vec(&self)?;
+        let public_key = base64::decode_config(&keys.public_key, base64::URL_SAFE_NO_PAD)?;
+        let public_key = RemoteKeyPairImpl::from_raw(&public_key);
+        let auth_secret = base64::decode_config(&keys.auth_secret, base64::URL_SAFE_NO_PAD)?;
+        let encrypted = Aes128GcmEceWebPushImpl::encrypt(
+            &public_key,
+            &auth_secret,
+            &bytes,
+            WebPushParams::default(),
+        )?;
+        let encrypted = base64::encode_config(&encrypted, base64::URL_SAFE_NO_PAD);
+        Ok(EncryptedSendTabPayload { encrypted })
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct TabData {
+    pub title: String,
+    pub url: String,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct PrivateSendTabKeys {
+    public_key: Vec<u8>,
+    private_key: Vec<u8>,
+    auth_secret: Vec<u8>,
+}
+
+impl PrivateSendTabKeys {
+    pub fn from_random() -> Result<Self> {
+        let key_pair = LocalKeyPairImpl::generate_random()?;
+        let private_key = key_pair.to_raw();
+        let public_key = key_pair.pub_as_raw()?;
+        let mut auth_secret = vec![0u8; 16];
+        RNG.fill(&mut auth_secret)
+            .map_err(|_| ErrorKind::RngFailure)?;
+        Ok(Self {
+            public_key,
+            private_key,
+            auth_secret,
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct SendTabKeysPayload {
+    /// Hex encoded kid (kXCS).
+    kid: String,
+    /// Base 64 encoded IV.
+    #[serde(rename = "IV")]
+    iv: String,
+    /// Hex encoded hmac.
+    hmac: String,
+    /// Base 64 encoded ciphertext.
+    ciphertext: String,
+}
+
+impl SendTabKeysPayload {
+    fn decrypt(self, ksync: &[u8], kxcs: &[u8]) -> Result<PublicSendTabKeys> {
+        // Most of the code here is copied from `EncryptedBso::decrypt`:
+        // we can't use that method as-it because `EncryptedBso` forces
+        // a payload id to be specified, which in turns make the Firefox
+        // Desktop commands implementation angry.
+        if hex::decode(self.kid)? != kxcs {
```

Comment:
Should this be a constant time comparison?

Reply:
@rfk ^

Reply:
No, I don't believe it needs to be, `kxcs` is more of a plaintext identifier than a secret.